python/pyarrow/_exec_plan.pyx

```diff
@@ -0,0 +1,297 @@
+        c_join_type = CJoinType_RIGHT_SEMI
+        left_columns = []
+    elif join_type == "left anti":
+        c_join_type = CJoinType_LEFT_ANTI
+        right_columns = []
+    elif join_type == "right anti":
+        c_join_type = CJoinType_RIGHT_ANTI
+        left_columns = []
+    elif join_type == "inner":
+        c_join_type = CJoinType_INNER
+        right_columns = set(right_columns) - set(right_keys)
+    elif join_type == "left outer":
+        c_join_type = CJoinType_LEFT_OUTER
+        right_columns = set(right_columns) - set(right_keys)
+    elif join_type == "right outer":
+        c_join_type = CJoinType_RIGHT_OUTER
+        left_columns = set(left_columns) - set(left_keys)
+    elif join_type == "full outer":
+        c_join_type = CJoinType_FULL_OUTER
+    else:
+        raise ValueError("Unsupported join type")
+
+    # Turn the columns to vectors of FieldRefs
+    for colname in left_columns:
+        c_left_columns.push_back(CFieldRef(<c_string>tobytes(colname)))
+    for colname in right_columns:
+        c_right_columns.push_back(CFieldRef(<c_string>tobytes(colname)))
+
+    # Add the join node to the execplan
+    if deduplicate:
+        c_decl_plan.push_back(
+            CDeclaration(tobytes("hashjoin"), CHashJoinNodeOptions(
+                c_join_type, c_left_keys, c_right_keys,
+                c_left_columns, c_right_columns,
+                _true,
+                <c_string>tobytes(left_suffix or ""),
+                <c_string>tobytes(right_suffix or "")
+            ))
+        )
+        if join_type == "full outer":
+            # In case of full outer joins, the join operation will output all columns
+            # so that we can coalesce the keys and exclude duplicates in a subsequent projection.
+            left_columns_set = set(left_columns)
+            right_columns_set = set(right_columns)
+            # Where the right table columns start.
+            right_table_index = len(left_columns)
+            for idx, col in enumerate(left_columns + right_columns):
+                if idx < len(left_keys):
+                    # Include keys only once and coalesce left+right table keys.
+                    c_projected_col_names.push_back(tobytes(col))
+                    c_projections.push_back(Expression.unwrap(
+                        Expression._call("coalesce", [
+                            Expression._field(idx), Expression._field(
+                                right_table_index+idx)
+                        ])
+                    ))
+                elif right_table_index <= idx < right_table_index+len(right_keys):
+                    # Do not include right table keys. As they would lead to duplicated keys.
+                    continue
+                else:
+                    # For all the other columns incude them as they are.
+                    # Just recompute the suffixes that the join produced as the projection
+                    # would lose them otherwise.
```

Comment:
It's not possible to already know the resulting column names (the schema) of the join operation before doing the projection? (so we could get the potentially suffixed column names from there by index)

Reply:
Not when relying on `CDeclaration` as the way to build the `ExecPlan`, I don't think saving those four lines of code values the amount of work that would be required to expose node types directly such that you can extract the `output_schema` of a plan that has not yet executed.

I'm all for creating a follow-up ticket if we care about changing that part, but my opinion, as expressed in https://ursalabs.zulipchat.com/#narrow/stream/180245-dev/topic/Preserve.20projection.20column.20names/near/276175594 is that the `ProjectNodeOptions` themselves should not force you to know the column names when you just want to preserve them.